controllers/stargate_controller.go

```diff
@@ -0,0 +1,340 @@
+	if err != nil && errors.IsNotFound(err) {
+
+		// FIXME is this correct?
+		cassdc := &cassdcapi.CassandraDatacenter{}
+		dcKey := client.ObjectKey{Namespace: req.Namespace, Name: stargate.Spec.DatacenterName}
+		err := r.Get(ctx, dcKey, cassdc)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				logger.Info("CassandraDatacenter resource not found. Ignoring since object must be deleted")
+				return ctrl.Result{}, nil
+			}
+			logger.Error(err, "Failed to get CassandraDatacenter")
+			return ctrl.Result{}, err
+		}
+
+		deployment := r.newStargateDeployment(stargate, cassdc)
+		logger.Info("Creating a new Stargate Deployment", "Deployment.Namespace", deployment.Namespace, "Deployment.Name", deployment.Name)
+		err = r.Create(ctx, deployment)
+		if err != nil {
+			logger.Error(err, "Failed to create new Stargate Deployment", "Deployment.Namespace", deployment.Namespace, "Deployment.Name", deployment.Name)
+			return ctrl.Result{}, err
+		}
+		// Deployment created successfully - return and requeue
+		return ctrl.Result{Requeue: true}, nil
+
+	} else if err != nil {
+		logger.Error(err, "Failed to get Stargate Deployment")
+		return ctrl.Result{}, err
+	}
+
+	// Ensure the deployment size is the same as the spec
+	size := stargate.Spec.Size
+	if *found.Spec.Replicas != size {
+		found.Spec.Replicas = &size
+		err = r.Update(ctx, found)
+		if err != nil {
+			logger.Error(err, "Failed to update Stargate Deployment", "Deployment.Namespace", found.Namespace, "Deployment.Name", found.Name)
+			return ctrl.Result{}, err
+		}
+		// Ask to requeue after 1 minute in order to give enough time for the
+		// pods be created on the cluster side and the operand be able
+		// to do the next update step accurately.
+		return ctrl.Result{RequeueAfter: time.Minute}, nil
+	}
+
+	// TODO Update the Stargate status
+
+	return ctrl.Result{}, nil
+}
+
+// newStargateDeployment returns a Deployment object for Stargate
+func (r *StargateReconciler) newStargateDeployment(stargate *api.Stargate, cassdc *cassdcapi.CassandraDatacenter) *appsv1.Deployment {
+
+	cassandraVersion := cassdc.Spec.ServerVersion
+
+	var clusterVersion string
+	if strings.HasPrefix(cassandraVersion, "3") {
+		clusterVersion = "3.11"
+	} else {
+		clusterVersion = "4.0"
+	}
+
+	var image string
+	var pullPolicy corev1.PullPolicy
+	containerImage := stargate.Spec.ContainerImage
+	if containerImage == nil {
+		if clusterVersion == "3.11" {
+			image = fmt.Sprintf("%s/%s:v%s", "stargateio", "stargate-3_11", stargate.Spec.ServerVersion)
+		} else {
+			image = fmt.Sprintf("%s/%s:v%s", "stargateio", "stargate-4_0", stargate.Spec.ServerVersion)
+		}
+		pullPolicy = corev1.PullIfNotPresent
+	} else {
+		image = fmt.Sprintf("%s/%s:%s", containerImage.Registry, containerImage.Repository, containerImage.Tag)
+		pullPolicy = containerImage.PullPolicy
+	}
+
+	var initImage string
+	var initPullPolicy corev1.PullPolicy
+	initContainerImage := stargate.Spec.InitContainerImage
```

Comment:
As I mentioned in an earlier comment, we won't need this init container. The Stargate controller can wait to create the Stargate Deployment until the CassandraDatacenter is ready.